### ISSUE TITLE
Request: Support HiDPI monitors

### ISSUE THREAD

**Issue by vanxining** [pkulchenko/ZeroBraneStudio#907]
Hi, thanks for this fantasy tool!
But it does not work well on HiDPI monitors:

![zb](https://user-images.githubusercontent.com/536291/39997140-72095d68-57b4-11e8-9cdc-9c39fa559ec7.png)

Some controls are very small. (The toolbar, tree control, etc.)
Is there any plan to enhance this?
Thanks again!

**Comment by mp035** [comment]
This is getting to the stage where the IDE is almost unusable on Linux HiDPI screens (dell xps15 shown, icons are impossible to see):
![screenshot from 2018-12-03 20-07-46](https://user-images.githubusercontent.com/9688384/49367351-8f0c2580-f737-11e8-90a3-712eb77a8540.png)

**Comment by pkulchenko** [comment]
@vanxining, @mp035, can you show what is being returned by `ide:GetMainFrame():GetContentScaleFactor()` when you execute it in the Console?

**Comment by pkulchenko** [comment]
@vanxining, @mp035, you may set a larger icon size for the toolbar, which will scale the available icons:. Try adding `toolbar.iconsize=32` or `toolbar.iconsize=48` to the config. It may look a bit ugly, but should make the icons more clickable. Even `toolbar.iconsize=64` should work.

I'll add some auto-scaling mechanism, but I need to add icons for the project tree and the search toolbar first (as there are no 24x24 sizes for those).

**Comment by mp035** [comment]
Thanks for the attention.  :-)
```
ide:GetMainFrame():GetContentScaleFactor()
1
```
From the docs (https://studio.zerobrane.com/doc-general-preferences):
> toolbar.iconsize = nil: set the size of the icons in the toolbar. **Only two sizes are currently supported: 16 and 24 pixels.** Starting from 1.11+ the default size for icons is 24 pixels on macOS or on screens 1500+ pixels wide; all the other configurations are using 16 pixel icons. You can still set the values to 16 or 24 as desired.

Probably needs an update to say that sizes other than 16 and 24 are now supported.

toolbar.iconsize=24 caused regular crashes when I tried it yesterday.

toolbar.iconsize=48 fixes the toolbar and makes it usable and doesn't seem to crash. 

 My only (minor) issue now is that the debug markers in the margin, and the icons in the file tree are still tiny.  Is there a setting for that?

Thanks again, and good work on zbstudio, it's a great application.

**Comment by pkulchenko** [comment]
> Probably needs an update to say that sizes other than 16 and 24 are now supported.

I agree; I'll update the documentation. BTW, what version of ZBS are you using?

> toolbar.iconsize=24 caused regular crashes when I tried it yesterday.

I wouldn't expect that and haven't seen anything similar. What version of Linux are you running and is it 32 or 64bit?

> My only (minor) issue now is that the debug markers in the margin, and the icons in the file tree are still tiny. Is there a setting for that?

Not yet, but there is going to be something to adjust that as well. Can you attach a screenshot on how the debug marker looks like? I'd expect this to be taken care of by the editor component itself, as it should be scaled along with the text in the editor.

**Comment by pkulchenko** [comment]
> toolbar.iconsize=24 caused regular crashes when I tried it yesterday.

@mp035, also, was there any additional information in the crash? Did it crash on launch or after some actions in the application?

**Comment by mp035** [comment]
it crashed immediately on launch (just got to see the window rendered for a split second).  It would crash 6 out of 7 times, and then start and run fine.

Screenshot as requested with breakpoint and execution line marker showing:
![screenshot from 2018-12-04 10-05-40](https://user-images.githubusercontent.com/9688384/49409465-81dd4e00-f7ac-11e8-99b8-fd733451d404.png)

**Comment by mp035** [comment]
Also an experiment (if it helps).  The debug markers definitely don't scale with text:

![screenshot from 2018-12-04 10-10-13](https://user-images.githubusercontent.com/9688384/49409570-f44e2e00-f7ac-11e8-8feb-af78895fbb90.png)

**Comment by pkulchenko** [comment]
no problem and thank you for the update!

I do see the issue with the breakpoint marker, although it's interested, as it definitely scales up and down (try to make it smaller than normal), it just seems to have some max value set, which is much smaller than it could be. You may try a slightly different marker, as it seems to scale better; try adding the following to the config: `styles.marker.breakpoint.ch = wxstc.wxSTC_MARK_ROUNDRECT`. (it should work for now, although I may replace `.ch` and `.st` for consistency with indicators and other styles)

**Comment by fsinisi90** [comment]
@pkulchenko Is there a way to change the breakpoint and bookmark icons? I don't see them on the icons folder, I guess those are some icons that came by default with wxwidgets, right? I want to customize them with my own icons. Many thanks in advance.

**Comment by pkulchenko** [comment]
@fsinisi90, yes, you should be able to do it using the Scintilla API for that (https://www.scintilla.org/ScintillaDoc.html#Markers) and then call `ide:AddMarker("breakpoint", markerid...)` on the marker you register, to register it with the IDE (and similar for the "bookmark"). There are `MarkerDefineBitmap` and `MarkerDefineRGBAImage` methods for the editor that you can use.

I haven't tested this, but don't see why it wouldn't work.

**Comment by fsinisi90** [comment]
Hmm, let's say I want to reuse the debugger icon from the toolbar. I've added these lines in `user.lua`:

```
local bitmap = ide:GetBitmap('DEBUG-BREAKPOINT-TOGGLE', 'TOOLBAR', wx.wxSize(16,16))
local marker = ide:GetEditor():MarkerDefineBitmap(0, bitmap)
```

And I'm getting this error:

`Error while processing configuration file: 'src/editor/package.lua:381: attempt to index field 'frame' (a nil value)'.`

What I'm doing wrong?

**Comment by pkulchenko** [comment]
@fsinisi90, it's because the environment is not fully set up at that time to execute those methods (there is no editor to use). You'll need to create an inline plugin to handle this. For example, add the following code to `user.lua`:

```
local function defineBreakpointBitmap(self, editor)
  local bitmap = ide:GetBitmap('DEBUG-BREAKPOINT-TOGGLE', 'TOOLBAR', wx.wxSize(16,16))
  editor:MarkerDefineBitmap(ide:GetMarker("breakpoint"), bitmap)
end
package {
  onEditorLoad = defineBreakpointBitmap,
  onEdiotrNew = defineBreakpointBitmap,
}
```

This sort of works, but it looks like the mask is no set correctly on the image, so I get a black box on windows (although your results may differ on macOS). If you are getting black boxes as well, I'll check what may be going on with the mask.

**Comment by fsinisi90** [comment]
It worked for me on Windows and Mac also. Many thanks!

On-topic now: Do you know if there is a way to add support for HiDPI screens? For what I've read here, I see that it's possible to use larger icons by adding a new `/res/48` folder and then setting `toolbar.iconsize=48`. But what I want is to create 48px icons and display them on 24px, using the capabilities of the retina display. Thanks again for the help, you rock! 🤘

**Comment by pkulchenko** [comment]
@fsinisi90, you may not need to do this. The version of wxwidgets that ships with ZBS is supposed to support hires images on macOS, so if you provide `DEBUG-BREAKPOINT-TOGGLE.png@2` image that is 48x48 in `res/24` directory it's supposed to work (this is just going by the documentation, as I haven't tested this).

**Comment by fsinisi90** [comment]
I've tried without success. I'll let you know if I can make it work, but apparently it's not working, at least on Mac (I'll try on a Windows with HiDPI later).

**Comment by pkulchenko** [comment]
@fsinisi90, it's not likely to work on windows, as the `@x` code only works on macOS as far as I can tell. There have been discussions to enable it in other contexts, but I don't remember if the actual implementation was submitted.

**Comment by pkulchenko** [comment]
> On-topic now: Do you know if there is a way to add support for HiDPI screens? For what I've read here, I see that it's possible to use larger icons by adding a new /res/48 folder and then setting toolbar.iconsize=48. But what I want is to create 48px icons and display them on 24px, using the capabilities of the retina display. Is that possible?

@fsinisi90, this should now be possible with the code available in all-binaires-190b branch. It takes content scale factor into account (`ide:GetContentScaleFactor()`) and will use 48x48 icons where 24px ones are requested (assuming content scale factor is set to 2). If the icons don't exist, it will scale 24px icons to match. This code should work much better on hidpi screens. You can still set the exact size you want using `toolbar.iconsize`, but the default values should work in most cases.

**Comment by pkulchenko** [comment]
@vanxining, @mp035, as reported earlier, the current beta version (https://github.com/pkulchenko/ZeroBraneStudio/tree/all-binaries-190c) makes several improvements for running on hidpi screens, so I'd be interested in your experience/feedback.

**Comment by mp035** [comment]
I just tried the mentioned beta version, and it runs perfectly.  Icons an breakpoint markers appear to scale fine.  I'd say you can close this issue.

**Comment by pkulchenko** [comment]
Thanks @mp035! I'll close it when the changes are merged into the master branch.

**Comment by fsinisi90** [comment]
> > On-topic now: Do you know if there is a way to add support for HiDPI screens? For what I've read here, I see that it's possible to use larger icons by adding a new /res/48 folder and then setting toolbar.iconsize=48. But what I want is to create 48px icons and display them on 24px, using the capabilities of the retina display. Is that possible?
> 
> @fsinisi90, this should now be possible with the code available in [all-binaires-190c branch](https://github.com/pkulchenko/ZeroBraneStudio/tree/all-binaries-190c). It takes content scale factor into account (`ide:GetContentScaleFactor()`) and will use 48x48 icons where 24px ones are requested (assuming content scale factor is set to 2). If the icons don't exist, it will scale 24px icons to match. This code should work much better on hidpi screens. You can still set the exact size you want using `toolbar.iconsize`, but the default values should work in most cases.

@pkulchenko How can I test it exactly? Should I have a 48px image called, for example `RUN.png@2` in the `res/24` folder or should I call it `RUN.png` and place it in a new `res/48` folder?

**Comment by pkulchenko** [comment]
> should I call it RUN.png and place it in a new res/48 folder?

This one. Since the actual requested image will be 48px (24px * 2 scale factor), it will be loaded from res/48 folder. If the image is missing, it should be scaled from 24px, so you can try with just one image in /48 folder.

**Comment by pkulchenko** [comment]
@fsinisi90, just one more note. If you do indeed have scale factor set to 2, you don't need to set `toolbar.iconsize=48`, as this will produce the wrong result. Just keep the default.

**Comment by pkulchenko** [comment]
This has been implemented as part of the changes in #999 that have been merged.
